### PR TITLE
Models refactor

### PR DESCRIPTION
**Pull request by RadoRado**

### CODE REVIEW

File: odin/authentication/services.py [hunk 1]
```diff
@@ -19,9 +19,22 @@ class Meta:
...
     class Meta:
         model = BaseUser
-        fields = ('id', 'email')
+        fields = (
+            'id',
+            'email',
+            'user_type'
+            )
+
+    def get_user_type(self, obj):
```
Comment (RadoRado) [inline odin/authentication/services.py]: I think it's a better idea to introduce a `get_user_type` service that does this calculation & have some constants - `STUDENT_TYPE` and `TEACHER_TYPE`

File: odin/education/apis/courses.py [hunk 2]
```diff
@@ -31,7 +53,15 @@ def get_students_count(self, obj):
     serializer_class = Serializer
 
     def get_queryset(self):
-        student = self.request.user.downcast(Student)
+        user = self.request.user
+
+        student = user.downcast(Student)
```
Comment (RadoRado) [inline odin/education/apis/courses.py]: we can move the downcasting to student after the `if` check - this way, if the check is `True`, we are going to do 1 downcast less

File: odin/education/apis/courses.py [hunk 3]
```diff
@@ -78,3 +109,71 @@ def get(self, request, course_id):
...
+
+    def post(self, request, course_id):
+        init_data = request.data
+        init_data['course'] = course_id
+        serializer = self.Serializer(data=init_data)
+        serializer.is_valid(raise_exception=True)
+        data = serializer.validated_data
+        code = data.pop('code')
+
+        included_task = create_included_task(**data)
```
Comment (RadoRado) [inline odin/education/apis/courses.py]: The entire business logic from `post` should be moved to a service

File: odin/education/apis/permissions.py [hunk 4]
```diff
@@ -17,6 +21,63 @@ def has_permission(self, request, view):
...
+
+
+class IsStudentOrTeacherInCoursePermission(BasePermission):
+    def has_permission(self, request, view):
+        user = request.user
+
+        student = user.downcast(Student)
+        teacher = user.downcast(Teacher)
+
+        course = get_object_or_404(view.get_queryset(), pk=view.kwargs.get('course_id'))
```
Comment (RadoRado) [inline odin/education/apis/permissions.py]: I think `course` should be attached to `view` before this permission is run. Something like `WithCourse` mixin

File: odin/education/apis/permissions.py [hunk 5]
```diff
@@ -17,6 +21,63 @@ def has_permission(self, request, view):
...
+
+        if (teacher and course.teachers.filter(id=teacher.id).exists())\
```
Comment (RadoRado) [inline odin/education/apis/permissions.py]: We can just return this boolean expression without the if/else statements

File: odin/education/management/commands/compare_solutions.py [hunk 6]
```diff
@@ -81,26 +81,24 @@ def handle(self, *args, **options):
...
-                    while passing_solutions:
-                        current_solution = passing_solutions.popleft()
-                        for next_solution in passing_solutions:
-                            if task.test.is_source():
-                                output = compare_code_solutions(current_solution, next_solution)
-                            else:
-                                output = compare_file_solutions(current_solution, next_solution)
-
-                            result += output
+            tasks = course.tasks.all()
```
Comment (RadoRado) [inline odin/education/management/commands/compare_solutions.py]: Not sure we need all of this

File: odin/education/migrations/0022_auto_20180323_1536.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
```
Comment (RadoRado) [inline odin/education/migrations/0022_auto_20180323_1536.py]: We should try squashing the migrations

File: odin/education/models.py [hunk 8]
```diff
@@ -156,41 +156,36 @@ class IncludedMaterial(BaseMaterial):
     material = models.ForeignKey(Material,
                                  on_delete=models.CASCADE,
                                  related_name='included_materials')
-    topic = models.ForeignKey('Topic',
-                              on_delete=models.CASCADE,
-                              related_name='materials')
+    week = models.ForeignKey('Week',
+                             related_name='materials',
+                             null=True,
+                             )
```
Comment (RadoRado) [inline odin/education/models.py]: Last bracket should be at the same indentation level as `week`

File: odin/education/models.py [hunk 9]
```diff
@@ -156,41 +156,36 @@ class IncludedMaterial(BaseMaterial):
...
+
+    course = models.ForeignKey(Course,
+                               related_name='included_materials',
+                               null=True,
+                               )
```
Comment (RadoRado) [inline odin/education/models.py]: Same here

File: odin/education/apis/permissions.py (hunk 4 above)
Comment (yaggul) [reply odin/education/apis/permissions.py]: Actually we didn't use that but i will keep it until you show me what you mean.

File: odin/education/models.py (hunk 8 above)
Comment (yaggul) [reply odin/education/models.py]: done

File: odin/education/models.py (hunk 9 above)
Comment (yaggul) [reply odin/education/models.py]: done

File: odin/authentication/services.py (hunk 1 above)
Comment (yaggul) [reply odin/authentication/services.py]: done

File: odin/education/apis/courses.py (hunk 2 above)
Comment (yaggul) [reply odin/education/apis/courses.py]: done

File: odin/education/apis/courses.py (hunk 3 above)
Comment (yaggul) [reply odin/education/apis/courses.py]: done

File: odin/authentication/services.py [hunk 10]
```diff
@@ -19,13 +19,28 @@ class Meta:
...
+
+
+def get_user_type(*, user: BaseUser) -> str:
+    STUDENT_TYPE = 'student'
+    TEACHER_TYPE = 'teacher'
+
+    if user.is_teacher():
+        return TEACHER_TYPE
+    elif user.is_student():
+        return STUDENT_TYPE
```
Comment (RadoRado) [inline odin/authentication/services.py]: raise a `TypeError` if we don't succeed with the ifs

File: odin/education/apis/courses.py [hunk 11]
```diff
@@ -31,7 +47,16 @@ def get_students_count(self, obj):
     serializer_class = Serializer
 
     def get_queryset(self):
-        student = self.request.user.downcast(Student)
+        user = self.request.user
+
+        teacher = user.downcast(Teacher)
+
+        if teacher:
```
Comment (RadoRado) [inline odin/education/apis/courses.py]: This is not correct. A user can be both teacher & student in different courses. We need either to return all courses (combine the queries) or split APIs

File: odin/education/management/commands/compare_solutions.py (hunk 6 above)
Comment (RadoRado) [reply odin/education/management/commands/compare_solutions.py]: Meaning - you can delete it

File: odin/education/models.py (hunk 8 above)
Comment (RadoRado) [reply odin/education/models.py]: It's not on the same indendation level.

Should be looking like this:

```python
week = models.ForeignKey(
    'Week',
    related_name='materials',
    null=True,
)
```

File: odin/education/models.py (hunk 9 above)
Comment (RadoRado) [reply odin/education/models.py]: Same as above

File: odin/authentication/tests/apis/test_change_password_api.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
+from test_plus import TestCase
+
+from django.test import Client
```
Comment (RadoRado) [inline odin/authentication/tests/apis/test_change_password_api.py]: Use either rest framework client OR test_plus client

**Pull request merged**